### ISSUE TITLE
question log 11/06

### ISSUE THREAD

**Issue by bladew** [bladew/Deep-Learning-for-Image-Denoising#1]
1. use sigmoid in the final net (2k-1 level net) or not
2. we need patch sampling method please add it into ImReader
3. pretraining seems expensive in computation. Is the procedure correct ?

**Comment by gji1** [comment]
1. I don't think we need the sigmoid, even though cropping the output pixel intensities to be in the range of [0,1] is a common trick for denoising.
2. I'm wondering if the patches are randomly sampled, or they just use all the overlapping patches in each image.
3. I thought the pre-training stage shouldn't be too slow since it only computes one pair of W and W' each time. We may discuss it in more details tomorrow.
4. Following 2, I'm wondering how they denoise an entire image after training, since they just trained on patches. Will the dimension match?

**Comment by bladew** [comment]
1. what's the patch size used?
2. how to reconstruct the whole image when the net is trained with patches?
3. follow the updating environment issue

**Comment by bladew** [comment]
1. the results are not good for da and sda (pre-training didn't help any). Need to check the the learning rate. And it's strange 2-layer sda has almost same result as da.
2. has modified the method to use patches but need a reconstruction method in util.py

**Comment by gji1** [comment]
1. We should see the training output (i.e. denoised patches) when you come back on Wednesday. If everything else is correct, then I think the main difference between our implementation and the paper lies in the regularization terms in the loss function. Is that correct? If yes, what regularization terms are you using right now?
2. The reconstruct method is added. Let me notice if you find anything wrong with it.

**Comment by bladew** [comment]
test on the reconstruct method (refer to the main part of util.py)
get result:
1.25637347748e-15
-2.88276347238e-14
-2.31941202355e-14
8.27931473379e-14
-1.15593298822e-14
-2.22686452611e-14
1.87089926884e-14
-1.2449069553e-13
1.7427682561e-13
-2.24022189688e-13
3.15969039127e-14
1.36041351795e-14
-3.60421328199e-15
2.33239642877e-13
-3.33387831231e-14
-8.87029165397e-14
-3.52219338765e-14
1.01441858386e-13
2.25480224769e-13
1.15520006755e-13
-9.28640844777e-14
-1.70824292212e-13
-1.02812723612e-14
-5.99303592863e-14
9.51218270817e-14
1.96121330981e-14
4.74652869092e-14
-4.64728082405e-14
-1.11553127846e-13
-1.64443111905e-13
2.36039486567e-14

Is this error caused by averaging over several patches or bugs in the method?

**Comment by bladew** [comment]
But anyway it's working now... We get some nice result from just 2 epochs.
0, PSNR change 14.1485 -> 19.7701
1, PSNR change 14.1607 -> 17.8315
2, PSNR change 14.1768 -> 21.6181
3, PSNR change 14.1234 -> 25.3496
4, PSNR change 14.156 -> 18.2499
5, PSNR change 14.1555 -> 20.306
6, PSNR change 14.1457 -> 21.8544
7, PSNR change 14.1606 -> 20.2148
8, PSNR change 14.166 -> 22.5478
9, PSNR change 14.1203 -> 24.4728

**Comment by bladew** [comment]
2-layer sda without pretraining, 1 epoch, produces worse result than da
2-layer sda with pretraining, 1epoch produces slightly better result than da (almost can't identify...)
I'll try 2 epochs for 2-layer sda with pretraining/ without pretraining
For now I haven't added regularization terms.

**Comment by bladew** [comment]
To do pretraining, set the W of da to be square (not necessary but convenient)

**Comment by bladew** [comment]
Interesting finding: after reducing hidden layer from 100 to 36, we got obviously better result in da. I guess it's the same with sda.

**Comment by gji1** [comment]
That looks weird... Does the training error also drop?

If no, I think it's due to overfitting, i.e. the fact that we didn't add
the regularization terms.

If yes, it may still because of that, as well as other reasons like faster
convergence under smaller parameter space.

**Comment by bladew** [comment]
best result so far with a 2-layer sda on classic data set:
0, PSNR change 14.155 -> 23.2278
1, PSNR change 14.1432 -> 25.438
2, PSNR change 14.1488 -> 26.9909
3, PSNR change 14.1854 -> 24.9423
avg is 25.12
but seems hard to go further
